flink-runtime/src/main/java/org/apache/flink/runtime/io/network/ShuffleEnvironment.java

```diff
@@ -0,0 +1,163 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network;
+
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.taskexecutor.TaskExecutor;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * Interface for the implementation of shuffle service locally on task executor.
+ *
+ * <p>Input/Output interface of local shuffle service is based on memory {@link org.apache.flink.runtime.io.network.buffer.Buffer}s.
+ * The task can request next available memory buffers from created here {@link ResultPartitionWriter}s to write shuffle data
+ * and buffers from created here {@link InputGate}s to read it.
+ *
+ * <h2>Service lifecycle management.</h2>
+ *
+ * <p>The interface contains method's to manage the lifecycle of the local shuffle environment:
+ * <ol>
+ *     <li>{@code start} is called when the {@link TaskExecutor} is being started.</li>
+ *     <li>{@code shutdown} is called when the {@link TaskExecutor} is being stopped.</li>
+ * </ol>
+ *
+ * <h2>Shuffle Input/Output management.</h2>
+ *
+ * <h3>Result partition management.</h3>
+ *
+ * <p>The interface implements a factory of result partition writers for the task output: {@code createResultPartitionWriters}.
+ * The created writers are grouped per task and handed over to the task thread upon its startup.
+ * The task is responsible for the writers lifecycle from that moment.
+ *
+ * <p>Partitions are released in the following cases:
+ * <ol>
+ *     <li>{@link ResultPartitionWriter#fail(Throwable)} and {@link ResultPartitionWriter#close()} are called
+ *     if the production has failed.</li>
+ *     <li>{@link ResultPartitionWriter#finish()} and {@link ResultPartitionWriter#close()} are called
+ *     if the production is done. The actual release can take some time
+ *     if 'the end of consumption' confirmation is being awaited implicitly
+ *     or the partition is later released by {@code releasePartitions(Collection<ResultPartitionID>)}.</li>
+ *     <li>{@code releasePartitions(Collection<ResultPartitionID>)} is called outside of the task thread,
+ *     e.g. to manage the local resource lifecycle of external partitions which outlive the task production.</li>
+ * </ol>
+ * The partitions, which currently still occupy local resources, can be queried with {@code updatePartitionInfo}.
```

Comment:
queried -> updated

Reply:
actually the method is wrong :) should be `updatePartitionInfo` -> `getUnreleasedPartitions`